direct-hs/src/Web/Direct/Api.hs

```diff
@@ -198,3 +178,51 @@ subscribeNotification client = do
 
 sendAck :: RPC.Client -> R.MessageId -> IO ()
 sendAck rpcClient mid = RPC.reply rpcClient mid $ Right $ M.ObjectBool True
+
+----------------------------------------------------------------
+
+handleNotifyCreateMessage
+    :: Config -> Client -> Message -> MessageId -> TalkId -> UserId -> IO ()
+handleNotifyCreateMessage config client msg msgid tid uid = do
+    me <- getMe client
+    let myid = userId me
+    when (uid /= myid && uid /= 0) $ do
+        mchan     <- findChannel client (tid, Just uid)
+        Just user <- findUser uid client
+        Just room <- findTalkRoom tid client
+        case mchan of
+            Just chan -> dispatch chan msg msgid room user
+            Nothing   -> do
+                mchan' <- findChannel client (tid, Nothing)
+                case mchan' of
+                    Just chan' -> dispatch chan' msg msgid room user
+                    Nothing    -> directCreateMessageHandler
+                        config
+                        client
+                        (msg, msgid, room, user)
+
+handleNotifyDeleteTalk :: Client -> TalkId -> IO ()
+handleNotifyDeleteTalk client tid = do
+    -- Close channels for talk
+    let chanDB = clientChannels client
+    getChannels chanDB tid >>= mapM_ (haltChannel chanDB)
+    -- Remove talk
+    talks <- getTalkRooms client
+    setTalkRooms client $ filter ((tid /=) . talkId) talks
+
+handleNotifyDeleteTalker
+    :: Client -> DomainId -> TalkId -> [UserId] -> [UserId] -> IO ()
+handleNotifyDeleteTalker client _ tid uids leftUids = do
+    -- Close channels that has no users
+    let chanDB = clientChannels client
+    chans <- getChannels chanDB tid
+    forM_ chans $ \chan -> do
+        chanAcqs <- getChannelAcquaintances client chan
+        let newChanAcqUids = filter (`notElem` leftUids) $ map userId chanAcqs
+        when (null newChanAcqUids) $ haltChannel chanDB chan
+    -- Update talk users
+    talks <- getTalkRooms client
+    setTalkRooms client $ map updateTalkUserIds talks
```

Comment:
The only concern that `handleNotifyDeleteTalk` and `handleNotifyDeleteTalker` have in common: `setTalkRooms` and `haltChannel` are not executed atomically.
I'm not sure that can cause a severe problem, but can make `clientChannels` and `clientTalkRooms` in inconsistent state.

Reply:
According to the oral discussion with @kazu-yamamoto:

- It's safer to update `clientTalkRooms` before `clientChannels`: Prevent other (channel) threads from seeing inconsistent state.
- Prefer `atomicModifyIORef'` to `writeIORef` in most cases: The former's overhead is very small. And it's safer especially when several threads writes the `IORef`.

Reply:
Thank you for helpful suggestions.
I pushed updates.